ccmbioinfo/lfs
# CARDS-2256: PREMs Overwrite LENGTH_OF_STAY_DAYS with calculation

- Added configuration to LengthOfStayMapper: Overwrite existing length by default
- Changed length of stay logic to ignore time of day when calculating stay duration
- Re-added Discard-IP-ShortStay discard filter

If you encounter issues running generate_compose_yaml due to a tzlocal error, there is a [test branch here](https://github.com/data-team-uhn/cards/tree/misc-prems-length-of-stay-test) which may fix your issue.

Testing instructions
1. Build with `-Pdocker` and navigate to `compose-cluster`
2. Clean up any old CARDS docker images: `docker-compose down && docker-compose rm && docker volume prune -f && ./cleanup.sh`
3. Prepare docker with the following flags: `python3 generate_compose_yaml.py --mssql --cards_project cards4prems --oak_filesystem --dev_docker_image --composum --adminer`
4. Start docker: `docker-compose build && docker-compose up -d`
5. Nacigate to http://localhost:1435/ and sign in:
- System: `MS SQL (beta)`
- Server: `mssql`
- Username: `sa`
- Password: `testPassword_`
- Database: `master`
6. Select Import on the left bar
7. Upload `cards/compose-cluster/mssql/processorImportTest.sql` and execute. It should say `38 queries executed OK`
8. Login to CARDS as admin
9. Navigate to http://localhost:8080/Subjects.importClarity. A mostly empty page should show success and Data import started
10. Navigate to the daashboard http://localhost:8080/content.html/Questionnaires/User and check which of patients 34 through 36 were imported.

Expected result:
Patients 34 and 35 should not be imported due to a length of stay of less than 1 day (10 hours or 23 hours 58 minutes)
Patient 36 should be imported as an inpatient or long form inpatient due to an 11 minute visit crossing midnight

11. `docker-compose down && docker-compose rm && docker volume prune -f && ./cleanup.sh`
12. Repeat steps 3 through 8
13. Delete the configuration for LengthOfStayMapper using `http://localhost:8080/system/console/configMgr`
14. Repeat steps 9 and 10. Verify the results are the same as before due to default handling if configuration is not present.
15. `docker-compose down && docker-compose rm && docker volume prune -f && ./cleanup.sh`
16. Repeat steps 3 through 8
17. Modify the configuration for LengthOfStayMapper using `http://localhost:8080/system/console/configMgr` to disable overwrite
14. Repeat steps 9 and 10. Verify that all of patients 34-36 are present as imported default LENGTH_OF_STAY of 1 is accepted
